### PR TITLE
Adds the VAEEncoder to StableDiffusion

### PR DESCRIPTION
**Pull request by LukeWood**
TODOS:

- [x] port and upload weights
- [x] figure out if padding needs to be done to non-512x512 images

### CODE REVIEW

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/models/generative/stable_diffusion/vae_encoder.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from keras_cv.models.generative.stable_diffusion.__internal__.layers.resnet_block import (
+    ResnetBlock,
+)
+
+
+class VAEEncoder(keras.Sequential):
+    def __init__(self):
+        super().__init__(
+            [
+                PaddedConv2D(128, 3, padding=1),
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/vae_encoder.py:23]: Done

**Comment by LukeWood** [comment]
@fchollet added weight loading.  Any review comments?

**Comment by LukeWood** [comment]
Weird; the VAEEncoder weights are not loading:

```
>>> model = keras_cv.models.StableDiffusion()
Downloading data from https://huggingface.co/fchollet/stable-diffusion/blob/main/vae_encoder.h5
36499/36499 [==============================] - 0s 2us/step
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lukewood/workspace/keras-cv/keras_cv/models/generative/stable_diffusion/stable_diffusion.py", line 124, in __init__
    self.image_encoder.load_weights(image_encoder_weights_fpath)
  File "/usr/local/lib/python3.9/site-packages/keras/utils/traceback_utils.py", line 67, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/usr/local/lib/python3.9/site-packages/h5py/_hl/files.py", line 424, in __init__
    fid = make_fid(name, mode, userblock_size,
  File "/usr/local/lib/python3.9/site-packages/h5py/_hl/files.py", line 190, in make_fid
    fid = h5f.open(name, flags, fapl=fapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 96, in h5py.h5f.open
OSError: Unable to open file (file signature not found)
```

I've tried deleting and re-downloading but still get the error.

**Comment by bhack** [comment]
> ttps://huggingface.co/fchollet/stable-diffusion/blob/main/

Have you verified `SHA256: 7ae08f918d4e3a686807474706a6821ca1ff3c71b84d98ba33081e6cb7aab9e3`?

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 2]
```diff
@@ -107,9 +113,15 @@ def __init__(self, img_height=512, img_width=512, jit_compile=False):
             origin="https://huggingface.co/fchollet/stable-diffusion/resolve/main/kcv_decoder.h5",
             file_hash="ad350a65cc8bc4a80c8103367e039a3329b4231c2469a1093869a345f55b1962",
         )
+        image_encoder_weights_fpath = keras.utils.get_file(
+            origin="https://huggingface.co/fchollet/stable-diffusion/blob/main/vae_encoder.h5",
+            file_hash="005390320d6095734833d0b442d11f413aae42c4d42872f3b352f7f53b5da122",
```
Comment (bhack) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:118]: it is now SHA256: 7ae08f918d4e3a686807474706a6821ca1ff3c71b84d98ba33081e6cb7aab9e3

Comment (bhack) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:118]: > OSError: Unable to open file (file signature not found)
The error message has a very low usability.

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:118]: Trying now with a manually specified hash algorithm.

Comment (bhack) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py]: https://huggingface.co/fchollet/stable-diffusion/blob/main/vae_encoder.h5
immagine

Comment (bhack) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py]: It doesn't match locally with wget:

```bash
sha256sum vae_encoder.h5 
c60fb220a40d090e0f86a6ab4c312d113e115c87c40ff75d11ffcf380aab7ebb`

md5sum vae_encoder.h5 
d3c5e2b8e4eee794417f9552fe0cb388  vae_encoder.h5
```

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py]: Updated to work.  try clearing your python cache

**Comment by LukeWood** [comment]
ok @fchollet this now works!

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 2 above)
Comment (bhack) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py]: > Updated to work. try clearing your "~/.keras" cache

It is not the cache, It was commited `c60` 14 minutes ago..

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py]: I know.  I just mean't try re-running now to confirm it works.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 3]
```diff
@@ -89,10 +90,15 @@ def __init__(self, img_height=512, img_width=512, jit_compile=False):
         self.text_encoder = TextEncoder(MAX_PROMPT_LENGTH)
         self.diffusion_model = DiffusionModel(img_height, img_width, MAX_PROMPT_LENGTH)
         self.decoder = Decoder(img_height, img_width)
+
+        # TODO(lukewood): should we support variable size image inputs here?
+        self.image_encoder = VAEEncoder()
```
Comment (fchollet) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:95]: This model will only be used for some use cases but creating it here will induce time and RAM overhead for all users. We should switch to a lazy model creation setup imo, something like `create_image_encoder()` called at the start of the method that requires it

File: keras_cv/models/generative/stable_diffusion/vae_encoder.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from keras_cv.models.generative.stable_diffusion.__internal__.layers.resnet_block import (
+    ResnetBlock,
+)
+
+
+class VAEEncoder(keras.Sequential):
+    def __init__(self):
+        super().__init__(
+            [
+                keras.layers.Input((512, 512, 3)),
```
Comment (fchollet) [inline keras_cv/models/generative/stable_diffusion/vae_encoder.py:35]: Please make the image dimensions keyword arguments, passed to be the user-provided values when creating the model

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:95]: we'll likely just use the image encoder directly in the tutorial.  Perhaps a lazy getter would work?

Re: `ImageEncoder` SGTM.

**Comment by LukeWood** [comment]
@fchollet what do you think about exposing the sub-models as APIs?  This would make writing TextualInversion a lot easier.  The one issue is the text models are very much NLP models.

**Comment by LukeWood** [comment]
Addressed all comments, PTAL.  My plan for now is to host the auxiliary modules needed in a 3p repo for now to get the guide written, then we can look at the best way to expose these APIs in a more permanent manner.

**Comment by LukeWood** [comment]
@fchollet let me know if you see any issues here.  I'm planning a pass on the Textual Inversion tutorial this weekend

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 5]
```diff
@@ -77,7 +82,14 @@ class StableDiffusion:
     - [Original implementation](https://github.com/CompVis/stable-diffusion)
     """
 
-    def __init__(self, img_height=512, img_width=512, jit_compile=False):
+    def __init__(
+        self,
+        img_height=512,
+        img_width=512,
+        input_img_height=512,
+        input_img_width=512,
```
Comment (ianstenbit) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:90]: I don't love stacking these arguments here, but I also think asking for an image encoder in the constructor is a bit weird.

I suppose we could add a method `addImageEncoder` or `withImageEncoder`, but that feels a bit awkward also.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 6]
```diff
@@ -89,10 +101,16 @@ def __init__(self, img_height=512, img_width=512, jit_compile=False):
...
+
+        self.input_img_height = input_img_height
+        self.input_img_width = input_img_width
+        self._image_encoder = None
+
         if jit_compile:
             self.text_encoder.compile(jit_compile=True)
             self.diffusion_model.compile(jit_compile=True)
             self.decoder.compile(jit_compile=True)
+            self.image_encoder.compile(jit_compile=True)
```
Comment (ianstenbit) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:113]: If I understand correctly, this causes us to instantiate an image encoder model whenever jit_compile is being used (regardless of whether it's actually going to be used). I think this is probably not the behavior we want.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 7]
```diff
@@ -281,6 +300,14 @@ def _get_unconditional_context(self):
 
         return unconditional_context
 
+    @property
+    def image_encoder(self):
```
Comment (ianstenbit) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:304]: I think if we're adding an image_encoder property, we may also want an API surface to use it (e.g. `encode_image`)?

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 5 above)
Comment (ianstenbit) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:90]: How would a user initialize StableDiffusion with custom pre-trained weights for the VAE?

File: keras_cv/models/generative/stable_diffusion/image_encoder.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    GroupNormalization,
+)
+from keras_cv.models.generative.stable_diffusion.__internal__.layers.padded_conv2d import (
+    PaddedConv2D,
+)
+from keras_cv.models.generative.stable_diffusion.__internal__.layers.resnet_block import (
+    ResnetBlock,
+)
+
+
```
Comment (ianstenbit) [inline keras_cv/models/generative/stable_diffusion/image_encoder.py:30]: I think a docstring for this encoder would probably be useful (at least to explain that it's a VAE and a reference)

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:90]: @fchollet any thoughts on this?  I also dont love stacking the arguments like this.  Perhaps we just assume the input & output images are of static shape?

Regarding custom weights: you don't.  StableDiffusion is only mean't to replicate exact numerics of the checkpoint; nothing more, nothing less.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:113]: Oh yes, that is correct.  Good catch!

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 7 above)
Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:305]: I was thinking that you just run `stable_diffusion.image_encoder.predict()`.  

Perhaps this is not a great API surface though?

Comment (fchollet) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:305]: Seems fine!

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 5 above)
Comment (fchollet) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:90]: Why not use the same size across the board?

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py]: I guess theres no use case YET to use a dynamic input shape; but perhaps there may be one at some point?  

In any case, I have modified this.

**Comment by LukeWood** [comment]
ok @fchollet @ianstenbit I think we're ready to merge!

File: keras_cv/models/generative/stable_diffusion/__internal__/layers/attention_block.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
```
Comment (ianstenbit) [inline keras_cv/models/generative/stable_diffusion/__internal__/layers/attention_block.py:1]: KerasCV authors

File: keras_cv/models/generative/stable_diffusion/__internal__/layers/resnet_block.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+from keras_cv.models.generative.stable_diffusion.__internal__.layers.group_normalization import (
+    GroupNormalization,
+)
+from keras_cv.models.generative.stable_diffusion.__internal__.layers.padded_conv2d import (
+    PaddedConv2D,
+)
+
+
+class ResnetBlock(keras.layers.Layer):
```
Comment (ianstenbit) [inline keras_cv/models/generative/stable_diffusion/__internal__/layers/resnet_block.py:25]: nit: should this be ResNetBlock?

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 11]
```diff
@@ -281,6 +293,14 @@ def _get_unconditional_context(self):
...
```
Comment (ianstenbit) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:297]: Maybe worth a docstring pointing to the ImageEncoder class or showing an example of its use?

**Review by ianstenbit — APPROVED** [review]

File: keras_cv/models/generative/stable_diffusion/__internal__/layers/resnet_block.py (hunk 10 above)
Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/__internal__/layers/resnet_block.py:25]: Lets leave this as a potential cleanup - this PR just moves the block but didn't add it.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 11 above)
Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:297]: Good point - added this!

**Comment by LukeWood** [comment]
Thanks @ianstenbit and @fchollet for the review, excited to use this in the TextualInversion guide!

**Review by fchollet — COMMENTED** [review]
LGTM

**Pull request merged**